EIPS/eip-165.md

```diff
@@ -0,0 +1,231 @@
+## Preamble
+
+    EIP: <to be assigned>
+    Title: Pseudo-Introspection, or standard interface detection
+    Author: Christian Reitwießner @chriseth, Nick Johnson @Arachnid, RJ Catalano @VoR0220, Fabian Vogelsteller @frozeman, Hudson Jameson @Souptacular, Jordi Baylina @jbaylina, Griff Green @griffgreen
+    Type: Standard Track
+    Category ERC
+    Status: Draft
+    Created: 2017-05-26
+
+## Simple Summary
+
+Creates a standard method to publish and discover what interfaces a smart contract implements.
+
+## Abstract
+
+This standard consists of the standardization of the following procedures:
+
+1. How a contract will publish the interfaces it implements.
+
+2. How one contract can know if another contract implements a standard interface.
+
+3. A generic method to determine if a contract implements this standard.
+
+4. A unique standard contract in the blockchain where you can ask if a contract implements an interface or not. Creating this contract has two main advantages:
+
+    4.1. It caches the answers, so it saves gas.
+    4.2. Many interfaces can be queried in the same call.
+
+## Motivation
+
+For some "standard interfaces" like the ERC 20 token interface ( [#20](https://github.com/ethereum/EIPs/issues/20) ), it is sometimes useful to query whether a contract supports the interface and if yes, which version of the interface, in order to adapt the way in which the contract is to be interfaced with. Specifically for #20, a version identifier has already been proposed. This proposal wants to standarize the concept of interfaces and map interface versions to interface identifiers.
+
+## Specification
+##### Standard interface identifier.
+
+Any standard interface will be identified by a specific `bytes4` number. This number will be the XOR of all `bytes4` of each method calculated as solidity does.
+
+For example, to calculate the id of a standard interface
+
+totalSupply()               -> Method ID: 0x18160ddd
+balanceOf(address)          -> Method ID: 0x70a08231
+transfer(address, uint256)  -> Method ID: 0xa9059cbb
+
+InterfaceID = 0x18160ddd ^ 0x70a08231 ^ 0xa9059cbb = 0xc1b31357
+
+##### EIP165 Interface
+
+Any contract that wants to implement a standard interface must also implement this method:
+
+    function supportsInterface(bytes4 interfaceID) returns (bool);
+
+This function must return:
+    `true` when `interfaceID` is 0x01ffc9a7 (EIP165 interface)
+    `false` when `interfaceID` is 0xffffffff
+    `true` for any `interfaceID` that this contract implements.
+    `false` for any other `interfaceID`
+
+##### Procedure to determine if a contract implements EIP165 standard.
+
+1. The source contact makes a `CALL` to the destination address with input data: 0x01ffc9a701ffc9a7 value: 0 and gas 30000
+2. If the call fails or return false, the destination contract does not implement EIP165
+3. If the call returns true, a second call is made with input data: 0x01ffc9a7ffffffff
```

Comment:
A typical example could be to ask if a wallet implements `ITokenFallback`.  In most cases, this call will `throw` or return false. The `return true` case, I see it very unprovable now, but after metropolis I see the possibility that some day solidity could be able to return values in the fallback function. (This could be very useful for implementing proxy contracts for example). I also do not control other languages like serpent or direct assembly contracts. So because of this possibility and to be sure there is no edge cases, I implemented the "Double call", But if you don't see it probable, we can remove this double check and go straight to point 3 checking for return false and exception.

Reply:
Personally I don't think it's necessary.